### DISCUSSION TITLE
Missing Version in OS Field - ??

### DISCUSSION THREAD

**Discussion by nvysyaraju-sift** [nielsbasjes/yauaa#1178] [Q&A]
Hello,

I see that for certain type of user agent strings the OS name is correctly detected but the version is not detected and getting `??` (Unknown) as returned value. But we do see the versions in the UA string, wanted to check if its possible to have parser detect the correct versions for these?

Here are few examples for correct detection of OS but unknown version (`??`)

For Android (OS) and ?? (Version)

```
Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Mobile Safari/537.36 bs=90 ts=0 ja3=ce980b4d231f964bc20e0adf4df13346 as=6167 - Version 10

com.kraken.invest.app/1.55.0.809421 (ANG-AN00 Android 12)

uphold-sdk-javascript/v2.4.1 (MT6737; 9; es_US) Wallet Android/5.1.1 - Version 9

uphold-sdk-javascript/v2.4.1 (exynos9820; 12; hu) Wallet Android/5.9.1 - Version 12

Android Course Hero/4.1.5 (vivo V2055; 13; en) - Version 13

Shopify Mobile/Android/9.129.0 (Build 33470 with API 31 on OPPO CPH2223)
```
For iOS (OS) and ?? (Version)

```
"uphold-sdk-javascript/v2.4.1 (iPhone13,4; 16.1.2; en) Wallet iOS/5.6.0" - Version 16 
Vestiaire Collective/5.75.0 (iPhone iOS:15.6.1 Scale:3.0)

Meetup-iOS/9.0.12 iPhone/16.3.1

Shopify Mobile/iOS/9.124.2 (iPhone15,3/com.jadedpixel.shopify/16.6.0) - Build 1184

myAuAppWebview/Mozilla/5.0 (iPhone; CPU iPhone OS 16_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148/iPhone10,2/auCSWebView/8.9.1/202309141100

com.reverb.mobile/5.11 (build 202309131643; like Reverb iOS) iOS/17.0.1 (iPhone14,6)

"Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) _LBANK_LBANK/iOS/5.0.1/iPhone 12/16.0/2_LBANK_"
```

Most of these UA strings may not be in the regular UA format, like starting with `Mozilla` but we do see the versions of OS for these strings so wanted to check here if its possible to have these detected?

Thank you.

**Comment by nielsbasjes** [comment]
I'll have a look. 
Most of these look like a random mess and it seems some have been modified (the `- Version 9` part).

Few questions I need an answer for before I can build anything:
- Is the ` - Version 9` and `- Version 10` at the end also part of the real useragent? I suspect it is not.
- Looking at `uphold-sdk-javascript/v2.4.1 (MT6737; 9; es_US) Wallet Android/5.1.1 - Version 9`: What is the Android version? Is it `5.1.1` or `9` or Unknown ?
- Looking at `_LBANK_LBANK/iOS/5.0.1/iPhone 12/16.0/2_LBANK_` I have no clue what is what. Here it becomes completely guessing; mainly because the logical ordering the versions do not make anysense and seem wrong. So perhaps LBANK App 5.0.1 on iPhone 12 with iOS 16.0 ??

**Reply by nvysyaraju-sift** [reply]
Answering above questions:
- Yes `- Version 9` and `- Version 10` are not part of the useragent string
- `uphold-sdk-javascript/v2.4.1 (MT6737; 9; es_US) Wallet Android/5.1.1`, I think it is `9` as `5.1.1 ` would not be currently supported as it is pretty old.
- I think you are right, these does not seem be with a proper order but 5.0.1 looks like the app version, 16.0 looks like the iOS version

Overall, most of these are useragents with app details, these strings probably look different for each source but is it be possible to have version detected for these or is it too tricky?

**Reply by nielsbasjes** [reply]
For some of these I can add support if I have either a spec or enough examples to make a good guess what the parts mean.
For some (like the lbank example) I have not enough confidence and examples to base anything on.

Also: Can you please provide me the **unmodified** useragents?

**Comment by nielsbasjes** [comment]
I have managed to classify some of the examples you provided.

Note that the first Android example you provided you are wrong about thinking this is Android 10 because this is an example of this: https://yauaa.basjes.nl/developer/reportingissues/#frozen-android-version .

**Reply by nvysyaraju-sift** [reply]
Thanks for sharing the `frozen-android-version` use case

Apologies for the delay to reply on previous comment, I am getting required permissions before I can share more examples of these useragents publicly. Thanks for adding some support to the already shared cases, will try it out online through parser-ui

**Reply by nvysyaraju-sift** [reply]
Hello @nielsbasjes 

Getting back to this particular discussion. Following up on the frozen OS version cases where the UA string might say Windows NT 10.0 / Android 10 but the actual versions could be manipulated right? Would there be some cases where the parser returns the versions as ?? even if the actual versions are given or unmanipulated in the string? I wanted to test some examples in the online parser to try and share the examples but it seems to be down today.

**Reply by nielsbasjes** [reply]
If there is no version or there is a reason to believe the version is nonsense then it should give ??
When there is a version it should show that.
If you find examples where it goes wrong then please report these.

Niels

**Reply by nvysyaraju-sift** [reply]
Thank you for the response. I have this particular example that shows the OS and version as `Windows NT ??`. 
Here is the UA string. 
`Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36`.

Also, may I know if this is the right link to the online parser (https://try.yauaa.basjes.nl/)?

**Reply by nielsbasjes** [reply]
Yes in this example the ?? is correct.
Yes that is the right link (I had a server problem today).

**Reply by nvysyaraju-sift** [reply]
> Yes in this example the ?? is correct.

Thank you for verifying it. Just curious, how can we say that this is manipulated and the version is not available and not 10.0? 

> Yes that is the right link (I had a server problem today).

👍 Now I see the page is up.
